cdap-app-fabric/src/main/java/co/cask/cdap/route/store/ZKRouteStore.java

```diff
@@ -0,0 +1,214 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.route.store;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.io.Codec;
+import co.cask.cdap.common.service.ServiceDiscoverable;
+import co.cask.cdap.common.zookeeper.ZKExtOperations;
+import co.cask.cdap.internal.guava.reflect.TypeToken;
+import co.cask.cdap.proto.id.ProgramId;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.SettableFuture;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import org.apache.twill.common.Threads;
+import org.apache.twill.zookeeper.NodeData;
+import org.apache.twill.zookeeper.OperationFuture;
+import org.apache.twill.zookeeper.ZKClient;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * RouteStore where the routes are stored in a ZK persistent node. This is intended for use in distributed mode.
+ */
+public class ZKRouteStore implements RouteStore {
+  private static final Logger LOG = LoggerFactory.getLogger(ZKRouteStore.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_INTEGER_TYPE = new TypeToken<Map<String, Integer>>() { }.getType();
+  private static final int ZK_TIMEOUT_SECS = 5;
+
+  private final ZKClient zkClient;
+  private final ConcurrentMap<ProgramId, SettableFuture<RouteConfig>> routeConfigMap;
+
+  @Inject
+  public ZKRouteStore(ZKClient zkClient) {
+    this.zkClient = zkClient;
+    this.routeConfigMap = Maps.newConcurrentMap();
+  }
+
+  @Override
+  public void store(final ProgramId serviceId, final RouteConfig routeConfig) {
+    Supplier<RouteConfig> supplier = Suppliers.ofInstance(routeConfig);
+    final SettableFuture<RouteConfig> settableFuture = SettableFuture.create();
+    Futures.addCallback(ZKExtOperations.createOrSet(zkClient, getZKPath(serviceId), supplier, ROUTE_CONFIG_CODEC, 10),
```

Comment:
Do we need to add a callback here? Since we are just waiting on a future, we could wait on the future returned by `ZKExtOperations.createOrSet()`, right?